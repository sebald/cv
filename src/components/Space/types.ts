export type SpaceProps = {
  className?: string;
};
