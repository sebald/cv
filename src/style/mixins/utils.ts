import { CSSProperties } from 'react';
import { StyledMixin } from '../theme';

export const propOrEmpty = <P extends keyof CSSProperties>(
  name: P
): StyledMixin<Pick<CSSProperties, P>> => (props: Pick<CSSProperties, P>) => {
  const p = props[name];
  return p ? { [name]: p } : {};
};
