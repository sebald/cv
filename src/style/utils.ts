import { theme } from './theme';

const { htmlFontSize } = theme;

export const px = (value: number) => `${value}px`;

export const pxToRem = (pixel: number, base = htmlFontSize) =>
  `${parseFloat((pixel / base).toFixed(5))}rem`;

export const remToPx = (rem: string, base = htmlFontSize) =>
  parseFloat(rem) * base;

export const ensurePx = (value: number | string, base = htmlFontSize) =>
  typeof value === 'string' ? remToPx(value, base) : value;
