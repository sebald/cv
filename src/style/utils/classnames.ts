export const mergeClassNames = (...names: (string | undefined)[]) => ({
  className: names
    .filter(x => typeof x === 'string')
    .join(' ')
    .trim(),
});
