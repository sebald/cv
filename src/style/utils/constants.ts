export const htmlFontSize = 16;

export const ratio = {
  perfectFourth: 1.333,
  perfectFifth: 1.5,
  goldenRatio: 1.618,
  majorSixth: 1.667,
};
