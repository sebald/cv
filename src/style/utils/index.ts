export * from './classnames';
export * from './constants';
export * from './createMixinComponent';
export * from './interpolation';
export * from './unit';
export * from './typography';
