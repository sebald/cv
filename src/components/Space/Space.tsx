/**
 * Stolen from:
 * https://github.com/jxnblk/grid-styled/tree/master/styled-space
 */
import * as React from 'react';
import { space, styled } from '../../style';

export type StyledProps = {
  className?: string;
};

export type SpaceProps = StyledProps & {};

export class StyledChildren extends React.Component<SpaceProps> {
  render() {
    const { className, children } = this.props;
    const styledChildren = React.Children.toArray(children).map(
      child =>
        React.isValidElement<any>(child)
          ? React.cloneElement(child, {
              className: [child.props.className || '', className!]
                .join(' ')
                .trim(),
            })
          : child
    );

    return <>{styledChildren}</>;
  }
}

export const Space = styled(StyledChildren)`
  ${space};
`;
