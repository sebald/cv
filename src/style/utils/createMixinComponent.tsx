import React from 'react';

import { styled } from '../styled-components';
import { StyledMixin } from '../theme';

import { mergeClassNames } from './classnames';
import { WithClassName } from './types';

/**
 * Idea stolen from:
 * https://github.com/jxnblk/grid-styled/tree/master/styled-space
 */
const StyledChildren: React.FC<WithClassName> = ({ className, children }) => {
  const styledChildren = React.Children.toArray(children).map((child) =>
    React.isValidElement<any>(child)
      ? React.cloneElement(
          child,
          mergeClassNames(child.props.className, className)
        )
      : child
  );

  return <>{styledChildren}</>;
};

export const createMixinComponent = <P extends {}>(
  ...mixins: StyledMixin<P>[]
) => styled(StyledChildren)([] as any, ...mixins);
