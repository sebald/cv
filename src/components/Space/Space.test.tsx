import * as React from 'react';

import { mountWithTheme } from '../../test.utils';
import { Space } from './Space';

test('render correctly', () => {
  const wrapper = mountWithTheme(<Space />);
  expect(wrapper).toMatchSnapshot();
});

test('render children', () => {
  const wrapper = mountWithTheme(
    <Space>
      <div>Hello</div>
      <h2>there</h2>
    </Space>
  );
  expect(wrapper).toMatchSnapshot();
});

test('add "spacing" classNames to children', () => {
  const wrapper = mountWithTheme(
    <Space my={2} mx={1}>
      <div>Hello</div>
      <h2>there</h2>
    </Space>
  );
  const { className } = wrapper.find('div').props();
  expect(className).toBeTruthy();
  expect(wrapper.find('h2').props().className).toEqual(className);
});

test('merges with existing child classNames', () => {
  const wrapper = mountWithTheme(
    <Space mx={2}>
      <div className="beep">Hello</div>
      <h2>hi</h2>
    </Space>
  );
  expect(wrapper.find('div').props().className).toMatch('beep');
});
