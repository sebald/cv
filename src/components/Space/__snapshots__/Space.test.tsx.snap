// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render children 1`] = `
<Styled(StyledChildren)>
  <StyledChildren
    className=""
  >
    <div
      className=""
      key=".0"
    >
      Hello
    </div>
    <h2
      className=""
      key=".1"
    >
      there
    </h2>
  </StyledChildren>
</Styled(StyledChildren)>
`;

exports[`render correctly 1`] = `
<Styled(StyledChildren)>
  <StyledChildren
    className=""
  />
</Styled(StyledChildren)>
`;
